### PR TITLE
Prevent custom matcher is called twice without compound

### PR DESCRIPTION
**Pull request by AlecksJohannes**
Hey lovely folks :wave:, I hope you all are doing well 

This PR here is to address https://github.com/rspec/rspec/issues/142 which is to prevent a custom matcher from calling twice without compound like `.and` or `.or`

### CODE REVIEW

**Comment by JonRowe** [comment]
Thanks for starting to tackle this, the issue as I see it is somehow this method is being responded to by the matcher, this is what needs to be fixed `.matcher().matcher` should cause a natural `NoMethodError` e.g.

```
     NoMethodError:
       undefined method `eq' for #<RSpec::Matchers::BuiltIn::Eq:0x00000001111948f0 @expected="ok">

               expect(ok).to eq(ok).eq
```

File: spec/rspec/matchers/dsl_spec.rb [hunk 1]
```diff
@@ -479,20 +521,20 @@ def foo
     end
 
     it "allows helper methods to be defined with #define_method to have access to matcher parameters" do
-      matcher = new_matcher(:name, 3, 4) do |a, b|
+      matcher = new_matcher(:matcher, 3, 4) do |a, b|
```
Comment (JonRowe) [inline spec/rspec/matchers/dsl_spec.rb:524]: Please revert these

**Comment by AlecksJohannes** [comment]
> Thanks for starting to tackle this, the issue as I see it is somehow this method is being responded to by the matcher, this is what needs to be fixed `.matcher().matcher` should cause a natural `NoMethodError` e.g.
> 
> ```
>      NoMethodError:
>        undefined method `eq' for #<RSpec::Matchers::BuiltIn::Eq:0x00000001111948f0 @expected="ok">
> 
>                expect(ok).to eq(ok).eq
> ```

Hello @JonRowe  thank you, sure I can raise `NoMethodError` rather than using `RSpec.warning` WDYT ?

**Comment by JonRowe** [comment]
> Hello @JonRowe thank you, sure I can raise NoMethodError rather than using RSpec.warning. WDYT ?

The fix should be to find where the method is being responded to and prevent it, I suspect its probably including a module it shouldn't do.

**Comment by AlecksJohannes** [comment]
> > Hello @JonRowe thank you, sure I can raise NoMethodError rather than using RSpec.warning. WDYT ?
> 
> The fix should be to find where the method is being responded to and prevent it, I suspect its probably including a module it shouldn't do.

Hey there JonRowe, sorry I have updated my comment above.

**Comment by AlecksJohannes** [comment]
I'll paste it here just to be sure 

I believe I understand what you said and I agree with you too. I tried to do that but I am not sure about removing or adjusting this piece of code, I played around with our code before and I can make it to work exactly like you said but we need to deal with 👇 

```rb
def respond_to_missing?(method, include_private=false)
   super || @matcher_execution_context.respond_to?(method, include_private)
end

def method_missing(method, *args, &block)
   if @matcher_execution_context.respond_to?(method)
      @matcher_execution_context.__send__ method, *args, &block
   else
       super(method, *args, &block)
   end
end
```

Otherwise, even calling `matcher().matcher` it will still work even though the method does not exist. Because of this blocker,  I had to provide another solution

P/S: If we were somehow to remove that `method_missing` I think you won't be able to call that custom matcher recursively which I think resulting in a breaking changes.

**Comment by pirj** [comment]
What concerns me the most, why are matchers available as methods on an i stance of a matcher in the first place?

**Comment by JonRowe** [comment]
Migrated to the monorepo as rspec/rspec#148

**Pull request closed without merging**